lib/extensions/require_nested.rb

```diff
@@ -6,23 +6,19 @@ def require_nested(name)
     filename = "#{self}::#{name}".underscore
     filename = name.to_s.underscore if self == Object
     if Rails.application.config.cache_classes
+      raise LoadError, "No such file to load -- #{filename}" unless ActiveSupport::Dependencies.search_for_file(filename)
       autoload name, filename
     else
       require_dependency filename
     end
 
+    # If the nested constant is defined, its top level equivalent must be
+    # defined too... otherwise the Rails autoloader can get confused.
+    # This is done by triggering required_nested at the top level.
```

Comment:
After reading this the first time it sounds like we're asserting that a top level equivalent must exist so I kinda ignored it.

Really this is saying that *if* there is a top level equivalent then we have to define it here otherwise <insert bad things> will happen.

Reply:
Yeha, I took this text from the original PR's comments...let me try to come up with something clearer...

Reply:
How about...

```
If the nested constant has a top-level constant with the same name, then both
must be defined at the same time, otherwise the Rails autoloader can get
confused.

For example, suppose we have the following

    # app/models/thing_grouper.rb
    class ThingGrouper < ApplicationRecord
      require_nested :Thing
    end

    # app/models/thing_grouper/thing.rb
    class ThingGrouper::Thing; end

    # app/models/thing.rb
    class Thing < ApplicationRecord; end

When the require_nested call is made, we will define `ThingGrouper`'s nested
`Thing`, but at the same time we must also define `::Thing` in order to allow
the Rails autloader to work correctly.  We do this by using a special-case call
to `require_nested` on `Object`, if a top-level `thing.rb` file exists.
```